examples/blame_example.ipynb

```diff
@@ -67,212 +61,246 @@
+    "            polygons[i] = origin + VR\n",
+    "\n",
+    "        #Check every combination of of agent polygons\n",
+    "        detected_collisions = []\n",
+    "        polygons = [Polygon([p[0], p[1], p[2], p[3]]) for p in polygons]\n",
+    "        for j in range(num_agents):\n",
+    "            for k in range(j+1,num_agents):\n",
+    "                try:\n",
+    "                    if polygons[j].intersection(polygons[k]).area:\n",
+    "                        detected_collisions.append((j,k))\n",
+    "                except GEOSException as e:\n",
+    "                    print(f\"Collision candidates {j} and {k} failed with error {e}.\")\n",
+    "                    pass\n",
+    "        \n",
+    "        #Check for collisions and acquire: agent IDs, collision start time, collision end time\n",
+    "        #Collect all collisions in the simulation\n",
+    "        for agent_tuple in detected_collisions:\n",
+    "            #If a colliding pair is detected and are not in the dictionary, add them here to track them\n",
+    "            if not agent_tuple in collisions_ongoing:\n",
+    "                collisions_ongoing[agent_tuple] = LogCollision(\n",
+    "                    collision_agents=agent_tuple,\n",
+    "                    start_time=t,\n",
+    "                    end_time=None\n",
+    "                )\n",
+    "        keys_to_delete = []\n",
+    "        for agent_tuple, collision in collisions_ongoing.items():\n",
+    "            if not collision.collision_agents in detected_collisions:\n",
+    "                #If a collision is no longer detected within the dictionary, assume the agents\n",
+    "                #are no longer colliding and the end time is the previous time step (when the collision was last observed)\n",
+    "                collisions_ongoing[agent_tuple].end_time = t-1\n",
+    "                collisions_all.append(collisions_ongoing[agent_tuple])\n",
+    "                keys_to_delete.append(agent_tuple)\n",
+    "            elif t >= SIMULATION_LENGTH-1: #The simulation has ended and the collision is ongoing\n",
+    "                #The collision has not necessarily ended at this time step but it is the last time step it was observed to occur\n",
+    "                collisions_ongoing[agent_tuple].end_time = t\n",
+    "                collisions_all.append(collisions_ongoing[agent_tuple])\n",
+    "                keys_to_delete.append(agent_tuple)\n",
+    "        #Only continue tracking collisions that have been observed in the last time step\n",
+    "        collisions_ongoing = {k:v for k, v in collisions_ongoing.items() if not k in keys_to_delete}\n",
+    "    \n",
+    "    return collisions_all\n",
     "\n",
     "# Simulate with `initialize`, `drive` and `light` until there are collisions.\n",
-    "while collisions is None:\n",
+    "for _ in range(20): #Attempt 20 simulations looking for a collision\n",
```

Comment:
Averagely, how long does it take to generate a collision?

Reply:
I haven't captured any data in that regards but anecdotally I've never see it go beyond 3 attempts. I just didn't want to leave the possibility of an infinite loop if the end user plays around with some settings that makes collisions unlikely.